tests/test_workflow.py

```diff
@@ -0,0 +1,270 @@
+                name="base_signal", fn=GoodDefnBase.base_signal
+            ),
+        },
+        queries={
+            "query1": workflow._QueryDefinition(name="query1", fn=GoodDefn.query1),
+            "query-custom": workflow._QueryDefinition(
+                name="query-custom", fn=GoodDefn.query2
+            ),
+            None: workflow._QueryDefinition(name=None, fn=GoodDefn.query3),
+            "base_query": workflow._QueryDefinition(
+                name="base_query", fn=GoodDefnBase.base_query
+            ),
+        },
+    )
+
+
+class BadDefnBase:
+    @workflow.signal
+    def base_signal(self):
+        pass
+
+    @workflow.query
+    def base_query(self):
+        pass
+
+
+class BadDefn(BadDefnBase):
+    # Intentionally missing @workflow.run
+
+    @workflow.signal
+    def signal1(self):
+        pass
+
+    @workflow.signal(name="signal1")
+    def signal2(self):
+        pass
+
+    @workflow.signal(dynamic=True)
+    def signal3(self, name: str, *args: Any):
+        pass
+
+    @workflow.signal(dynamic=True)
+    def signal4(self, name: str, *args: Any):
+        pass
+
+    # Intentionally missing decorator
+    def base_signal(self):
+        pass
+
+    @workflow.query
+    def query1(self):
+        pass
+
+    @workflow.query(name="query1")
+    def query2(self):
+        pass
+
+    @workflow.query(dynamic=True)
+    def query3(self, name: str, *args: Any):
+        pass
+
+    @workflow.query(dynamic=True)
+    def query4(self, name: str, *args: Any):
+        pass
+
+    # Intentionally missing decorator
+    def base_query(self):
+        pass
+
+
+def test_workflow_defn_bad():
+    with pytest.raises(ValueError) as err:
+        workflow.defn(BadDefn)
+
+    assert "Invalid workflow class for 7 reasons" in str(err.value)
+    assert "Missing @workflow.run method" in str(err.value)
+    assert (
+        "Multiple signal methods found for signal1 (at least on signal2 and signal1)"
+        in str(err.value)
+    )
+    assert (
+        "Multiple signal methods found for <dynamic> (at least on signal4 and signal3)"
+        in str(err.value)
+    )
+    assert (
+        "@workflow.signal defined on BadDefnBase.base_signal but not on the override"
+        in str(err.value)
+    )
+    assert (
+        "Multiple query methods found for query1 (at least on query2 and query1)"
+        in str(err.value)
+    )
+    assert (
+        "Multiple query methods found for <dynamic> (at least on query4 and query3)"
+        in str(err.value)
+    )
+    assert (
+        "@workflow.query defined on BadDefnBase.base_query but not on the override"
+        in str(err.value)
+    )
+
+
+def test_workflow_defn_local_class():
+    with pytest.raises(ValueError) as err:
+
+        @workflow.defn
+        class LocalClass:
+            @workflow.run
+            async def run(self):
+                pass
+
+    assert "Local classes unsupported" in str(err.value)
```

Comment:
Don't remember why this isn't supported

Reply:
During my sandboxing research, it became clear I need to reference the workflow class by name across the boundary since I want to instantiate the workflow in a completely different interpreter/runtime/environment. Local classes are not cached/imported/referenceable that way.

This was a choice by me just to avoid having to concern myself with it and test for it only to have it probably taken away in phase 3 anyways. I also can't imagine a benefit of doing so, so I just prevent it for now.

If there's a reasonable reason to support it we can.

Reply:
That's okay especially it's required for sandboxing.
Haven't read the full error string, I'd just mention the reason there in case it's missing.

Reply:
I am hesitant to put our reasons for disallowing something on the error message. "Local classes unsupported because the class needs to be referenceable by name", "Arg and args are mutually exclusive because it'd be confusing if we allowed both", "Must put @workflow.defn on the class not just the base class because we think it's better to be more explicit about which workflows are which and it's less ambiguous to derive the name", etc.

I can definitely add a comment where I throw though to clarify our position.

Reply:
Comment added